draft-ion.adoc

```diff
@@ -208,79 +200,176 @@ represented as follows:
+
+If a member name is not equal to the case-sensitive octet sequence `meta`, an Ion parser MUST NOT identify a member as a _Meta Object_.
+
+The member names within an identified _Meta Object_ MUST be unique; Ion parsers MUST either
+reject _Meta Object_ members with duplicate member names or use a JSON
+parser that returns only the lexically last duplicate member, as specified
 in {uri-ecmaScript-5-1-15-12}[Section 15.12 (The JSON Object)] of {uri-ecmaScript-5-1}[ECMAScript 5.1].
 
-The set of fields that a _Meta Object_ must contain to be considered valid
+The set of members that a _Meta Object_ must contain to be considered valid
 is resource-dependent.  In the absense of any resource-specific requirements,
-any field not understood by implementations MUST be ignored.
+any member not understood by implementations MUST be ignored.
+
+[[links]]
+== Links
+
+An Ion Link is an Ion Object where its `meta` member contains an {uri-IRIs}[IRI] member named `href`. An Ion Link is therefore JSON serialization of a {uri-rfc5988}[web link] to another resource.
+
+Ion parsers MUST identify any JSON structure matching this definition as an Ion Link.  Anything that does not match this defined structure MUST NOT be considered an Ion Link.
+
+A non-normative example of a minimal Ion link:
+
+.Example simplest Ion link:
+[source,json]
+----
+{ "meta": { "href": "https://api.example.com/corporations/acme" } }
+----
+
+=== Link Relation Types
+
+All Ion links have one or more <<links, link relation types>>.  A link's total set of link relation types is the set union of the link's implicit link relation type and the link's explicit link relation types.
+
+==== Implicit Link Relation Type
+
+All Ion links have exactly one implicit link relation type.  The implicit link relation type can be one of three possible values depending on where the link is located in the Ion content:
+
+1. Root Object
+2. Named Link
+3. Collection Resource `items` array element
+
+===== Root Object Implicit Link Relation Type
+
+If the Root Object is also found to be an Ion Link, it has an implicit link relation type of `self`.
+
+A non-normative example of a Root Object that is also an Ion Link:
+
+[source,json]
+----
+{
+  "meta": { "href": "https://api.example.com/users/1" },
+  "name": "Joe"
+}
+----
 
-=== Registered Metadata Field Names
+===== Named Link Implicit Link Relation Type
 
-The following field names are registered in the IANA Ion Metadata field registry
-defined in [TBD].  None of the fields defined below are intended to be mandatory
+If a JSON Object member value is an Ion Link, the Ion Link is implicitly assigned a link relation type that equals the member name.
+
+A non-normative example:
+
+[source,json]
+----
+{
+  "employer": { "meta": { "href": "https://api.example.com/corporations/acme" } }
+}
+----
+
+The above example shows an Ion Link member named `employer`.  The octet sequence `employer` is therefore assigned as the link's implicit link relation type.
```

Comment:
This, to me, feels like a lot to do for supporting pluralized member names for arrays. I also feels this could hinder evolvability of the document.

Let's take this customer representation for example.

```json
{
  "name": "John Doe",
  "order": {
    "meta": {
      "href": "/order/1"
    }
  }
}
```

Here the implied link relation is `order` for the Ion object. In the future, though, if I decide to evolve this to allow for multiple orders, the representation changes.

```json
{
  "name": "John Doe",
  "orders": [
    {
      "meta": {
        "rel": ["order"],
        "href": "/order/1"
      }
    },
    {
      "meta": {
        "rel": ["order"],
        "href": "/order/2"
      }
    }
  ]
}
```

This means that (a) the server has to change how it builds the representation because it has a different name and (b) the client has to look for the `order` links in a different location.

Now if I want to use this new Collection Resource, it changes things even further.

```json
{
  "name": "John Doe",
  "orders": {
    "meta": {
      "itemRel": ["order"]
    },
    "items": [
      {
        "meta": {
          "href": "/order/1"
        }
      },
      {
        "meta": {
          "href": "/order/2"
        }
      }
    ]
  }
}
```

To me, this feels like a lot of rules, and the only thing I see coming from it is you can pluralize the property name that has an array as the value.

Going back to the original example, if you allow the names to be implicit link relations, you resolve all of this and allow for removing the Collection Resource section from the spec (because `item` would act as the implicit link relations). It also allows for evolvability of the resource because the order link relation hasn't moved or changed names.

```json
{
  "name": "John Doe",
  "order": [
    {
      "meta": {
        "href": "/order/1"
      }
    },
    {
      "meta": {
        "href": "/order/2"
      }
    }
  ]
}
```

Lastly, another option to resolve this would be to not use implicit link relations and require that a link MUST have a `rel`. This would allow you to use whatever property name you wanted and the client/server introspects the objects rather than relying on the property name.

---

I'm not arguing for any of these, just want to give some thoughts on the tradeoffs I see. HAL for instance solves this by propagating the link relation to each item in the array, so you would have singular implicit names. I know these questions will come up as people start using this spec as I've seen it in the HAL discussion group. Would be good to work them out now :)